sql/src/main/java/io/crate/executor/transport/ChainableActions.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.executor.transport;
+
+import com.google.common.collect.ImmutableList;
+import io.crate.exceptions.Exceptions;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+public class ChainableActions {
+
+    /**
+     *  Runs the given list of chainable actions sequentially, if one fails, all previous actions will be rolled back
+     *  by calling the {@link ChainableAction#undo()} callable of each action in reverse order.
+     */
+    @SuppressWarnings("ConstantConditions")
+    public static <ReturnType> CompletableFuture<ReturnType> run(List<? extends ChainableAction<ReturnType>> actions) {
+        assert actions.size() > 0 : "Empty list of ChainableActions";
+
+        List<ChainableAction<ReturnType>> previousActions = new ArrayList<>(actions.size());
+        ChainableAction<ReturnType> lastAction = actions.get(0);
+        CompletableFuture<ReturnType> future = lastAction.doIt();
+
+        for (int i = 1; i < actions.size(); i++) {
+            ChainableAction<ReturnType> action = actions.get(i);
+            previousActions.add(lastAction);
+            lastAction = action;
+            future = future
+                .thenApply(r -> new Result<>(r, null))
```

Comment:
I think `.thenApply().exceptionally()` could be replaced with `.whenComplete((r, t) -> new Result(r, t))`, saving one intermediate new future instance